theano/sandbox/cuda/cusolver.py

```diff
@@ -37,116 +39,109 @@ class GpuSolve(GpuOp):
+            # switch the "transpose" flag.
             if self.trans in ('T', 'C'):
                 trans = 'N'
             else:
                 trans = 'T'
 
-            # Convert b to F-order from c-order.
+            # Convert b to F-order from C-order.
             b_cpy = dimshuffle(b, (1, 0)).reshape((b.shape[0], b.shape[1]))
 
             # This copy forces allocation of a new C-contiguous buffer
             # and returns it.
             A_cpy = A.copy()
             b_cpy = b_cpy.copy()
 
-            def cusolver_gpu_solve(A_, b_, trans='T'):
-                A_shape = A_.shape
-                b_shape = b_.shape
-
-                assert(len(A_shape) == 2)
-                assert(len(b_shape) == 2)
-
-                if trans in ['T', 'C']:
-                    trans = 1
-                    l, n = A_shape
-                    k, m = b_shape
-                    if n != k:
-                        raise ValueError('A and b must be aligned.')
-                elif trans in ['N']:
-                    trans = 0
-                    n, l = A_shape
-                    k, m = b_shape
-                    if l != m:
-                        raise ValueError('A and b must be aligned.')
-                else:
-                    raise ValueError('Invalid value for trans')
-
-                lda = max(1, n)
-                ldb = max(1, n, l)
-
-                A_ptr = A_.gpudata
-                b_ptr = b_.gpudata
-
-                if cusolver_handle[0] is None:
-                    cusolver_handle[0] = cusolver.cusolverDnCreate()
-
-                workspace_size = cusolver.cusolverDnSgetrf_bufferSize(
-                        cusolver_handle[0], m, n, A_ptr, lda)
-                
-                if (thunk.workspace is None or 
-                    thunk.workspace.size != workspace_size):
-                    thunk.workspace = CudaNdarray.zeros((workspace_size,))
-                
-                if thunk.pivots is None or thunk.pivots.size != min(m, n):
-                    thunk.pivots = CudaNdarray.zeros((min(m, n),))
+            assert(len(A.shape) == 2)
+            assert(len(b.shape) == 2)
+
+            if trans in ['T', 'C']:
+                trans = 1
+                l, n = A.shape
+                k, m = b.shape
+                if n != k:
+                    raise ValueError('A and b must be aligned.')
+            elif trans in ['N']:
+                trans = 0
+                n, l = A.shape
+                k, m = b.shape
+                if l != m:
+                    raise ValueError('A and b must be aligned.')
+            else:
+                raise ValueError('Invalid value for trans')
+
+            lda = max(1, n)
+            ldb = max(1, n, l)
+
+            A_ptr = A_cpy.gpudata
+            b_ptr = b_cpy.gpudata
 
-                if thunk.dev_info is None:
-                    thunk.dev_info = CudaNdarray.zeros((1,))
+            if cusolver_handle is None:
```

Comment:
I would move the the init of the handle inside the make_thunk method, but outside the thunk() function. This will make profiling give more useful information by having it in the linker phase and not in the execute phase.